streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/BalanceSubtopologyGraphConstructor.java

```diff
@@ -0,0 +1,186 @@
+        }
+
+        for (int clientIndex = 0; clientIndex < clientList.size(); clientIndex++) {
+            final UUID processId = clientList.get(clientIndex);
+            final int capacity = originalAssignedTaskNumber.get(processId);
+            final int secondStageClientNodeId = getSecondStageClientNodeId(taskIdList, clientList, tasksForTopicGroup, clientIndex);
+            graph.addEdge(secondStageClientNodeId, sinkId, capacity, 0, 0);
+        }
+
+        graph.setSourceNode(SOURCE_ID);
+        graph.setSinkNode(sinkId);
+
+        // Run max flow algorithm to get a solution first
+        final long maxFlow = graph.calculateMaxFlow();
+        if (maxFlow != taskIdList.size()) {
+            throw new IllegalStateException("max flow calculated: " + maxFlow + " doesn't match taskSize: " + taskIdList.size());
+        }
+
+        return graph;
+    }
+
+    @Override
+    public boolean assignTaskFromMinCostFlow(final Graph<Integer> graph,
+        final List<UUID> clientList, final List<TaskId> taskIdList,
+        final Map<UUID, ClientState> clientStates,
+        final Map<UUID, Integer> originalAssignedTaskNumber, final Map<TaskId, UUID> taskClientMap,
+        final BiConsumer<ClientState, TaskId> assignTask,
+        final BiConsumer<ClientState, TaskId> unAssignTask,
+        final BiPredicate<ClientState, TaskId> hasAssignedTask) {
+
+        final SortedMap<Subtopology, Set<TaskId>> sortedTasksForTopicGroup = new TreeMap<>(tasksForTopicGroup);
+
+        int taskNodeId = 0;
+        int topicGroupIndex = 0;
+        int tasksAssigned = 0;
+        boolean taskMoved = false;
+        for (final Entry<Subtopology, Set<TaskId>> kv : sortedTasksForTopicGroup.entrySet()) {
+            final SortedSet<TaskId> taskIds = new TreeSet<>(kv.getValue());
+            for (final TaskId taskId : taskIds) {
+                final Map<Integer, Graph<Integer>.Edge> edges = graph.edges(taskNodeId);
+                for (final Graph<Integer>.Edge edge : edges.values()) {
+                    if (edge.flow > 0) {
+                        tasksAssigned++;
+                        final int clientIndex = getClientIndex(edge.destination, taskIdList, clientList, topicGroupIndex);
+                        final UUID processId = clientList.get(clientIndex);
+                        final UUID originalProcessId = taskClientMap.get(taskId);
+
+                        // Don't need to assign this task to other client
+                        if (processId.equals(originalProcessId)) {
+                            break;
+                        }
+
+                        unAssignTask.accept(clientStates.get(originalProcessId), taskId);
+                        assignTask.accept(clientStates.get(processId), taskId);
+                        taskMoved = true;
+                    }
+                }
+                taskNodeId++;
+            }
+            topicGroupIndex++;
+        }
+
+        // Validate task assigned
```

Comment:
Seems this validation code (starting here to the end of the method) is the same as for min-traffic? Should we extract a helper method and share code?